sui-execution/latest/sui-move-natives/src/crypto/group_ops.rs

```diff
@@ -751,3 +767,138 @@ pub fn internal_pairing(
+
+    let e_ref = pop_arg!(args, VectorRef);
+    let e = e_ref.as_bytes_ref();
+    let to_type = pop_arg!(args, u8);
+    let from_type = pop_arg!(args, u8);
+
+    let cost_params = &context
+        .extensions()
+        .get::<NativesCostTable>()
+        .group_ops_cost_params
+        .clone();
+
+    let result = match (Groups::from_u8(from_type), Groups::from_u8(to_type)) {
+        (Some(Groups::BLS12381G1Uncompressed), Some(Groups::BLS12381G1)) => {
+            native_charge_gas_early_exit_option!(
+                context,
+                cost_params.bls12381_g1_uncompressed_to_g1_cost
+            );
+            e.to_vec()
+                .try_into()
+                .map_err(|_| FastCryptoError::InvalidInput)
+                .map(bls::G1ElementUncompressed::from_trusted_byte_array)
+                .and_then(|e| bls::G1Element::try_from(&e))
+                .map(|e| e.to_byte_array().to_vec())
+        }
+        (Some(Groups::BLS12381G1), Some(Groups::BLS12381G1Uncompressed)) => {
+            native_charge_gas_early_exit_option!(
+                context,
+                cost_params.bls12381_g1_to_g1_uncompressed_cost
+            );
+            parse_trusted::<bls::G1Element, { bls::G1Element::BYTE_LENGTH }>(&e)
+                .map(|e| bls::G1ElementUncompressed::from(&e))
+                .map(|e| e.into_byte_array().to_vec())
+        }
+        _ => Err(FastCryptoError::InvalidInput),
+    };
+
+    match result {
+        Ok(bytes) => Ok(NativeResult::ok(cost, smallvec![Value::vector_u8(bytes)])),
+        // Since all Element<G> are validated on construction, this error should never happen unless the requested type is wrong.
+        Err(_) => Ok(NativeResult::err(cost, INVALID_INPUT_ERROR)),
+    }
+}
+
+/***************************************************************************************************
+ * native fun internal_sum_of_uncompressed
+ * Implementation of the Move native function `internal_sum_of_uncompressed(type:u8, terms: &vector<vector<u8>>): vector<u8>`
+ *   gas cost: group_ops_bls12381_g1_sum_of_uncompressed_base_cost + len(terms) * group_ops_bls12381_g1_sum_of_uncompressed_cost_per_term
+ **************************************************************************************************/
+pub fn internal_sum(
+    context: &mut NativeContext,
+    ty_args: Vec<Type>,
+    mut args: VecDeque<Value>,
+) -> PartialVMResult<NativeResult> {
+    debug_assert!(ty_args.is_empty());
+    debug_assert!(args.len() == 2);
+
+    let cost = context.gas_used();
+
+    if !(is_g1_uncompressed_supported(context)) {
+        return Ok(NativeResult::err(cost, NOT_SUPPORTED_ERROR));
+    }
+
+    let cost_params = &context
+        .extensions()
+        .get::<NativesCostTable>()
+        .group_ops_cost_params
+        .clone();
+
+    // The input is a reference to a vector of vector<u8>'s
+    let inputs = pop_arg!(args, VectorRef);
+    let group_type = pop_arg!(args, u8);
+
+    let length = inputs
```

Comment:
should we limit this using a protocol config like we did for msm? e.g., by the number of additions that costs roughly the time it takes to do one bls sig verification?

Reply:
Set to 2000 for now. This takes ~1ms to compute on my machine.